diesel/src/pg/expression/expression_methods.rs

```diff
@@ -1111,3 +1111,114 @@ impl InetOrCidr for Inet {}
+    /// Creates a PostgreSQL `||` operator
+    ///
+    /// This operator merges two `Jsonb` objects.
+    ///
+    /// # Example
+    ///
+    /// ```rust
+    /// # include!("../../doctest_setup.rs");
+    /// #
+    /// # table! {
+    /// #     characters {
+    /// #         id -> Integer,
+    /// #         a_b_c -> Jsonb,
+    /// #         d_e_f -> Jsonb,
+    /// #     }
+    /// # }
+    /// #
+    /// # #[cfg(feature = "serde_json")]
+    /// # fn main() {
+    /// #     run_test();
+    /// # }
+    /// #
+    /// # fn run_test() {
+    /// #     use diesel::insert_into;
+    /// #     use diesel::associations::HasTable;
+    /// #     use self::characters::dsl::*;
+    /// #     use serde_json::json;
+    /// #
+    /// #     let connection = connection_no_data();
+    /// #     diesel::sql_query("CREATE TABLE characters (
+    /// #         id SERIAL PRIMARY KEY,
+    /// #         a_b_c JSONB NOT NULL,
+    /// #         d_e_f JSONB NOT NULL
+    /// #     )").execute(&connection);
+    /// #
+    /// #     // insert into a_b_c
+    /// #     let a_b_c_data = json!({
+    /// #            "a": 1,
+    /// #            "b": 1,
+    /// #            "c": 1,
+    /// #            "letters": [
+    /// #                 "a",
+    /// #                 "b",
+    /// #                 "c",
+    /// #            ],
+    /// #     });
+    /// #     diesel::insert_into(a_b_c).values(a_b_c_data);
+    /// #
+    /// #     // insert into d_e_f
+    /// #     let d_e_f_data = json!({
+    /// #             "d": 1,
+    /// #             "e": 2,
+    /// #             "f": 3,
+    /// #             "letters": [
+    /// #                 "d",
+    /// #                 "e",
+    /// #                 "f",
+    /// #             ]
+    /// #     });
+    /// #     diesel::insert_into(d_e_f).values(d_e_f_data);
+    /// #
+    ///       let projected_merged_json = json!({
+    ///               "a": 1,
+    ///               "b": 2,
+    ///               "c": 3,
+    ///               "d": 1,
+    ///               "e": 2,
+    ///               "f": 3,
+    ///               "letters": [
+    ///                   "a",
+    ///                   "b",
+    ///                   "c",
+    ///                   "d",
+    ///                   "e",
+    ///                   "f",
+    ///               ]
+    ///     });
+    ///
+    ///       let merged_json = characters::table.select(characters::a_b_c.merge(characters::d_e_f)).load(&connection);
+    ///       assert_eq!(projected_merged_json, merged_json);
+    /// #
+    /// # }
+    /// #
+    /// # #[cfg(not(feature = "serde_json"))]
+    /// # fn main() {}
+    /// ```
+    fn merge<T: AsExpression<Self::SqlType>>(self, other: T) -> dsl::JsonbMerge<Self, T::Expression>
+    where
+        <Self as Expression>::SqlType: SqlType,
+    {
+        Grouped(JsonbMerge::<Self, T::Expression>::new(
+            self,
+            other.as_expression(),
+        ))
+    }
+}
```

Comment:
In order to allow handling cases where one is nullable and the other is not, we could do something similar to what is done with `And`:
https://github.com/diesel-rs/diesel/blob/9edd8acb51e0dcd44ccdbc7aae38c323005a3481/diesel/src/expression_methods/bool_expression_methods.rs#L41-L49
(That is, being generic on an additional ST and putting a bound on `JsonbMerge<Self, T::Expression>: Expression`.)